### DISCUSSION TITLE
Performance of createListenerMiddleware Predicate in Redux Toolkit

### DISCUSSION THREAD

**Discussion by guyeilon** [reduxjs/redux-toolkit#4605] [General]
Hi Redux Team,

I have a question regarding the `createListenerMiddleware` in Redux Toolkit. Specifically, I want to know how the `predicate` function behaves in terms of performance.

### My Question:
- Does the `predicate` function get evaluated after every action is dispatched, regardless of the action type or target?
- If so, could this cause performance issues in applications where actions are dispatched frequently, particularly if the `predicate` contains more complex logic (e.g., state comparisons or function calls)?
- Are there any best practices or recommendations to optimize the performance of `createListenerMiddleware` when using `predicate`?

I’m concerned about whether using `predicate` in a heavily action-driven environment could degrade performance, and I’d appreciate any insights or suggestions to avoid potential bottlenecks.

Thanks in advance for your help!

**Comment by markerikson** [comment]
Yes, it does get evaluated after _every_ action.

However, it shouldn't be a perf bottleneck. Predicates are supposed to be trivial checks that immediately return a boolean. That's even less work than all the selectors that will run after every action.

Now, yes, those _are_ more function calls, so there is an `O(n)` aspect to perf here - 100 predicate calls is literally more expensive than 50 or 10 or 1 predicate calls. But in practice these should have effectively no meaningful impact on perf, especially compared to rendering components or anything else.
